### PR TITLE
Refactor generators

### PR DESCRIPTION
**Pull request by dansbecker**
@sallamander I'm pretty unhappy with how this came out. Can you look at image_generator.py and let me know if you have any thoughts about how to simplify it.

I still think classes is the right way to do it.  Just think this implementation is ugly.

Also happy to talk through it by Google Hangout if that's easier.

On the upside, changing the output of `start_box` from 4 integers to a mask/arrayshould be really easy from here.  So, I'll do that as soon as we think the structure of the generator classes is more-or-less set.

### CODE REVIEW

File: motion_tracker/utils/crop_vis.py [hunk 1]
```diff
@@ -27,8 +27,8 @@ def show_stages_of_random_crop(img, box_coords, output_width=256, output_height=
                                                   output_width, output_height)
     show_img(final_img, final_box_coords.as_array())
 
-def show_img(img, box_coords=None, window_name="Happy Dance Image", 
-                      msec_to_show_for=1500):
+def show_img(img, box_coords=None, window_name="Happy Dance Image",
+                      msec_to_show_for=1500, coords_are_fracs=False):
     """Show an image, potentially with surrounding bounding boxes
```
Comment (sallamander) [inline motion_tracker/utils/crop_vis.py:32]: Out of curiosity, is this giving us the ability to basically plot without having to un-normalize the predictions after training?

File: motion_tracker/utils/image_generator.py [hunk 2]
```diff
@@ -4,212 +4,195 @@
...
+        """
+        self.crops_per_image = crops_per_image
+        self.batch_size = batch_size
+        self.output_width = output_width
+        self.output_height = output_height
+        self.desired_dim_ordering = desired_dim_ordering
+
+
+    def are_accumulators_full(self):
+        return self.X['start_img'].shape[0] >= self.batch_size
```
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:36]: I think common practice is to wrap something like this with the `@property`, and then to reference it like an attribute. I think you could then change it's name to `full`, or something like that.

File: motion_tracker/utils/image_generator.py [hunk 3]
```diff
@@ -4,212 +4,195 @@
...
+
+    def flow(self):
+        """The function yielding an iterator."""
+        while True:
+            self.reset_accumulators()
+            while not self.are_accumulators_full():
+                self.add_to_accumulators()
+            self.remove_accumulator_overflow()
```
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:44]: This is a question.. I thought we decided that we weren't going to care about this, and let the generator just keep going, since the keras model just takes whatever it is given? Is this in here just in case we decide we want it back?

File: motion_tracker/utils/image_generator.py [hunk 4]
```diff
@@ -4,212 +4,195 @@
...
+        empty_img_accumulator = np.zeros([0, self.output_width, self.output_height, 3]).astype('uint8')
+        empty_box_accumulator = np.zeros([0, 4]).astype('uint16')
+        empty_digit_accumulator = np.zeros([0]).astype('uint16')
+        self.X = {'start_img': empty_img_accumulator,
+                  'start_box': empty_box_accumulator,
+                  'end_img': empty_img_accumulator}
+        self.y = {'x0': empty_digit_accumulator,
+                  'y0': empty_digit_accumulator,
+                  'x1': empty_digit_accumulator,
+                  'y1': empty_digit_accumulator}
```
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:62]: I'm wondering at the possibility of just creating original copies of the `empty_img_accumulator`, `empty_box_accumulator`, and `empty_digit_accumulator`, and then just setting something called the `present_img_accumulator`, `present_box_accumulator`, and `present_digit_accumulator` as copies of these upon each reset. There are upsides and downsides I think of to doing it that way versus this current way, but I thought I'd throw it out there.

File: motion_tracker/utils/image_generator.py [hunk 5]
```diff
@@ -4,212 +4,195 @@
...
+            self.y[field] = self.y[field][:self.batch_size]
+
+    def ensure_dim_ordering(self):
+        if self.desired_dim_ordering == 'th': # swap because data comes in 'tf' format
+            self.X['start_img'] = self.X['start_img'].swapaxes(1, 3)
+            self.X['end_img'] = self.X['end_img'].swapaxes(1, 3)
+
+    def normalize(self, img):
+        return (img-100) / 60
+
```
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:83]: So, I'm not sold that this is a good idea, but since we'll only need to theoretically normalize the images once, we could do this outside of this class all together. Now, that would require that we normalize and then save the normalized images to disk, which is probably a pretty bad idea. But, it's an option and would remove some (potentially) extraneous code from this class. However, this also assumes that we're doing the same normalization across images from all batches, which may/may not be true. If we do image normalization across batches, then moving this outside isn't something we could do.

**Comment by sallamander** [comment]
@dansbecker Still going to look this over a couple of more times. All the comments I just made are  smallish kinds of things, rather than anything in terms of the overall structure. I'm trying to think of some way that we could possibly restructure it as well, wondering if we couldn't cut out one of the levels of inheritance somehow. So, I'm still going to look this over another couple of times and am not done, but for the time being I'll be pondering how to potentially restructure this. I typically benefit from some time to chew on things, so I think it's good that I looked at it now and can continue to think, even if I didn't actually propose anything that big here.

File: motion_tracker/utils/crop_vis.py (hunk 1 above)
Comment (dansbecker) [inline motion_tracker/utils/crop_vis.py]: yes.  I meant to add crop_vis to this PR.

Comment (dansbecker) [inline motion_tracker/utils/crop_vis.py:32]: Yes.

File: motion_tracker/utils/image_generator.py (hunk 3 above)
Comment (dansbecker) [inline motion_tracker/utils/image_generator.py:44]: I started to worry it would be the source of some subtle bug later on, when we assume we are working with a fixed batch size and it isn't true.  I'm sort of torn either way on this.

File: motion_tracker/utils/image_generator.py (hunk 4 above)
Comment (dansbecker) [inline motion_tracker/utils/image_generator.py:62]: Not sure I understand what you are thinking here.  Can you give some more detail about it (e.g. how it differs from the current approach)?

File: motion_tracker/utils/image_generator.py (hunk 5 above)
Comment (dansbecker) [inline motion_tracker/utils/image_generator.py:83]: My hunch has been that it doesn't matter one way or the other.  I'll remove it for now.

Comment (dansbecker) [inline motion_tracker/utils/image_generator.py:83]: It's also a very uncommon way to do normalization... normally people subtract off the average (over many images) of each pixel/channel value.  That potentially could matter, but is very close to what I actually did, since those values were all very close to each other.

File: motion_tracker/utils/crop_vis.py (hunk 1 above)
Comment (sallamander) [inline motion_tracker/utils/crop_vis.py:32]: 👍

File: motion_tracker/utils/image_generator.py (hunk 3 above)
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:44]: I think it might be better to just leave it in here, then.

File: motion_tracker/utils/image_generator.py (hunk 4 above)
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:62]: It wouldn't really be too different, more moving around creating of those arrays to the init and then using copies of them in this method. So, I could imagine doing something in the `__init__` where we do something like: 

```
self.empty_img_accumulator = ....
self.empty_box_accumulator = .....
self.empty_digit_accumulator = ....
```

Then, in this method we do: 

```
self.X = {'start_img': self.empty_img_accumulator.copy(),
                  'start_box': self.empty_box_accumulator.copy(),
                  'end_img': self.empty_img_accumulator.copy()}
self.y = {'x0': self.empty_digit_accumulator.copy(),
              y0': self.empty_digit_accumulator.copy(),
              'x1': self.empty_digit_accumulator.copy(),
              'y1': self.empty_digit_accumulator.copy()}
```

Now that I type this out, I like it much less. So, in reality I don't think this would really be beneficial to do.

File: motion_tracker/utils/image_generator.py (hunk 5 above)
Comment (sallamander) [inline motion_tracker/utils/image_generator.py:83]: Yeah, that all makes sense. Thought I'd throw it out there in case.

**Comment by sallamander** [comment]
Okay, so I think I have maybe two possibilities for how we might make this cleaner. 
1. Get rid of the middle layer of inheritance - the `CompositeGenerator`. This as well as the `ImagenetGenerator` and `AlovGenerator` all inherit from the same base class, and I don't see a huge point in having the `CompositeGenerator`. I might be missing the motivation for having that, so maybe we should discuss? When I thought of having this as classes, I imagined having one kind of base class (like the `ImagePairAndBoxGen`) that the `AlovGenerator` and `ImagenetGenerator` would inherit from, and then maybe add/alter a little of it's functionality. I think this is possible without the middle layer of the `CompositeGenerator`. 
2. What if we redid the `filtered_bb.csv` that gets read in in the `ImagenetGenerator` to match the format of the `AlovGenerator` (we could also do the inverse of that, but I think that's harder and maybe not possible to do). Then, we don't actually need these separate generators - we just need to pass in the right filepaths that point to the bounding box csv and the images. If you're open to this, I think this actually makes a good deal of sense, since we always expect that images will go into the net in the same format (e.g. a starting/ending image and corresponding bounding boxes).

**Comment by dansbecker** [comment]
@sallamander I've got it refactored as we previously discussed, and we are using a mask for the original bounding box rather than just integers.

Give it a look, but I think it is ready to merge. Error is still around 10% of the img width after it sees 10000 training observations. Also, running it with more filters in each convolutional layer may help, though we'll probably run into a RAM limit on my GPU if we do that.  If I have time, I'll set it running on the CPU over the weekend.

**Comment by sallamander** [comment]
I think this looks good, too. Two questions that I want to check: 
1. Did you plan on the joining of the `ImagenetGenerator` and `AlovGenerator` classes being a separate PR? I know you mentioned that, and think I misunderstood exactly what was going in that PR. So, just wanted to double check here to see if that was still the plan?
2. Are we punting on any further refactoring of the `add_to_accumulators` in `ImagePairAndBoxGen`? We discussed it, and I think were both enough on the fence about whether that would be worth it, so I'm good with that. Just wanted to check. 

Other than that, good stuff! Thanks for putting the generator classes together 👍 👊

**Comment by dansbecker** [comment]
In theory we could do the merging of those two classes, but I don't think it has such a big payoff that it matters either way. I think training this on a larger GPU with more filters in each level, and then saving the model, may be the next step.  That, and visualizing the crops, so we have an idea of how good this thing is at this point.
